src/librustc_resolve/lib.rs

```diff
@@ -3512,6 +3635,62 @@ fn path_names_to_string(path: &Path, depth: usize) -> String {
     names_to_string(&names[..])
 }
 
+/// When an entity with a given name is not available in scope, we search for
+/// entities with that name in all crates. This method allows outputting the
+/// results of this search in a programmer-friendly way
+fn show_candidates(session: &mut DiagnosticBuilder,
+                   span: &syntax::codemap::Span,
+                   candidates: &SuggestedCandidates) -> () {
+
+    // don't show more than 4 results, so we're
+    // consistent with the trait suggestions
+    const MAX_CANDIDATES: usize = 5;
+
+    // we want consistent results across executions, but candidates are produced
+    // by iterating through a hash map, so make sure they are ordered:
+    let paths = &candidates.candidate_paths;
+    let mut path_strings: Vec<_> = paths.into_iter()
+                                        .map(|p| path_names_to_string(&p, 0))
+                                        .collect();
+    path_strings.sort();
+
+    // behave differently based on how many candidates we have:
+    if !paths.is_empty() {
+        if paths.len() == 1 {
+            session.fileline_help(
+                *span,
+                &format!("You may be referring to `{}`, which you need \
+                to import with the `use` keyword.", &path_strings[0]),
+            );
+        } else {
+            session.fileline_help(
+                *span,
+                "You may be referring to one of: ",
+            );
+            let count = path_strings.len() - MAX_CANDIDATES + 1 as usize;
+
+            for (idx, path_string) in path_strings.iter().enumerate() {
+                if idx == MAX_CANDIDATES - 1 && count > 1 {
```

Comment:
... or remove `count` and change the second condition to `path_strings.len() > MAX_CANDIDATES`